### ISSUE TITLE
memory-inefficient area:id filter

### ISSUE THREAD

**Issue by tyrasd** [drolbr/Overpass-API#359]
Consider the following two almost identical queries:

    way["building"](area:3600285864);

and:

    area(3600285864);
    way["building"](area._);

The first one fails (with a *Query run out of memory using about 2048 MB of RAM* runtime error) while the second one works fine. Here's the [example](http://overpass-turbo.eu/s/kYG) in overpass turbo.

//cc @mocnik-science

**Comment by mmd-osm** [comment]
Thanks for reporting!

Preliminary analysis:

Case 1: `way["building"](area:3600285864);`

```C
bool Area_Constraint::delivers_data(Resource_Manager& rman)
{
  if (!area->areas_from_input())
return false;
```

Irrespective of the actual area size, delivers_data will be always false, leading to check_keys_late being `false` in query.cc

This has further impact on `filter_id_list` in https://github.com/drolbr/Overpass-API/blob/master/src/overpass_api/statements/query.cc#L157

Method may not stop processing after 2^20 entries because of check_keys_late being false. This way, the method accumulates around 210 Mio entries!

Case 2:

More sophisticated logic to determine area size in https://github.com/drolbr/Overpass-API/blob/master/src/overpass_api/statements/area_query.cc#L294-L308 

-> delivers_data is true. Early exit of `filter_id_list` is possible.

**Comment by drolbr** [comment]
Fixed in dba448fca45d48eff3a850eaa366f78a8355a69e
